clarify local and input variables (#12334)
hashicorp/packer: website/content/docs/templates/hcl_templates/locals.mdx

---

> page_title: Local Variables - HCL Configuration Language

Reviewer: I wonder if we should update this document to be consistent with the Terraform [local values documentation](https://developer.hashicorp.com/terraform/language/values/locals). What do you think?

Author: @nywilken Oh, so you'd rather see me go the other direction, away from "variables" towards "values" for locals? Is Terraform consistent in the usage of "values"?

I don't really see a problem with doing that. Might lead to less confusion ... put more distance between the concepts so folks realize they're quite different in behavior.

description: >-

>   Local variables assign a name to an expression that can then be used multiple

Reviewer 2: I'm not sure if descriptions should be active or passive voice so feel free to disregard if this is wrong:
```suggestion
  Local variables assign a name to an expression that you can use multiple
```

  times within a folder.
---

# Local Variables

`@include 'from-1.5/beta-hcl2-note.mdx'`

There are two kinds of variables in HCL Packer templates: Input variables,
sometimes simply called "variables", and Local variables, also known as
"locals". Input variables may have defaults, but those defaults can be
overridden using command line options, environment variables, or variable

> definitions files. Local variables cannot be overridden.

Reviewer 2: I believe we avoid contractions:
```suggestion
definitions files. Local variables can not be overridden.
```

This page is about local variables. To learn about input variables, see the
[input variables](/packer/docs/templates/hcl_templates/variables) page.

> Local variables assign a name to an expression, that can then be used multiple

Reviewer 2: ```suggestion
Local variables assign a name to an expression, which you can use multiple
```

times within a folder. The expression is evaluated at run time, and can
reference input variables, other local variables, data sources, and HCL
functions.

Input variable and local variable usage are introduced in the [_Variables
Guide_](/packer/guides/hcl/variables).

## Examples

> Local variables are defined in `local` or `locals` blocks:

Reviewer: ```suggestion
Local variables are defined in a `local` or `locals` block:
```

  name_prefix         = "${var.name_prefix != "" ? var.name_prefix : local.default_name_prefix}"
}

> # Local variables can be interpolated elsewhere using the "local." prefix.

Reviewer: Interpolation relates more to legacy JSON templates then it does HCL. 
```suggestion
# Local variables can be referenced using the "local." prefix.
```
